### PR TITLE
Migrate shared-buffer constructs to the new dependency runtime (`object()`, `tuple()`, `concat()`, `merge()`)

### PR DESCRIPTION
**Pull request by dahlia**
## Summary

This PR migrates all four shared-buffer constructs to the dependency runtime introduced in #752, replacing the old state-type inspection protocol with centralized runtime-based dependency resolution. This is the third step in the #750 redesign sequence (#751 → #752 → #753).

The old protocol required constructs to inspect parser states for `DeferredParseState`, `DependencySourceState`, and `PendingDependencySourceState` markers, then resolve dependencies through recursive state-tree traversal (`resolveDeferredParseStates`). The new path uses `DependencyRuntimeContext` as a shared resolution engine: constructs pre-complete dependency sources once, register values in the runtime, then resolve `DeferredParseState` objects against the populated runtime. The runtime is propagated to nested constructs via `ExecutionContext.dependencyRuntime`, so sources registered at any nesting depth are visible to derived parsers anywhere in the tree.

## Design

Each construct's `complete()` now follows a three-phase flow:

1. **Pre-complete dependency sources** via `preCompleteAndRegisterDependencies()`. This is the single evaluation point for lazy defaults: each source parser's `complete()` is called exactly once, its result is stored for reuse in phase 3, and the dependency value is registered in the runtime. The function handles all four cases from the old protocol (pending state in arrays, undefined state with pending `initialState`, `wrappedDependencySourceMarker`, and non-null state from `bindConfig`/`bindEnv` wrappers).

2. **Resolve deferred state** via `resolveStateWithRuntime()`. This performs a two-pass traversal of the state tree: first collecting all `DependencySourceState` values (which may override defaults with actual CLI-provided values), then replaying all `DeferredParseState` objects using the populated runtime. The whole-tree approach ensures resolution is order-independent regardless of field declaration order.

3. **Complete each field** using the resolved state. Fields that were pre-completed in phase 1 reuse the stored result to avoid re-evaluating non-idempotent lazy defaults. Other fields are completed normally, with `unwrapCompleteResult()` handling `DependencySourceState` returns from the old protocol.

The suggest paths (*suggestObjectSync*/*suggestObjectAsync*, merge suggest, *buildSuggestRegistry*) use a lighter approach: `collectExplicitSourceValues()` for metadata-based source detection, `collectSourcesFromState()` for state-tree harvesting, and `completeDependencySourceDefaults()` for lazy default evaluation. Suggest paths deliberately do not replay `DeferredParseState`, since tab completion only needs the dependency registry populated, not the actual replay results. This avoids triggering async `parseWithDependency` calls whose promises would be dropped in the sync resolver.

## Old-protocol bridge

Not all wrappers compose `dependencyMetadata` yet. In particular, `bindConfig()` and `bindEnv()` do not propagate dependency metadata from their inner parsers. To handle this during the transition period, the construct-level pre-completion step uses old-protocol markers (`wrappedDependencySourceMarker`, `isPendingDependencySourceState`) to detect dependency sources. This bridge will be removed once all wrappers compose metadata (#755).

For cases 1-3 (pending/missing state), pre-completion only registers the result when `complete()` returns an actual `DependencySourceState`. This is important because wrappers like `map()` intentionally break the dependency chain, and force-wrapping their plain results would incorrectly register transformed values as source values. Case 4 (`bindConfig`/`bindEnv`) uses `wrapAsDependencySourceState()` to force-wrap, since these wrappers return plain results by design.

## Non-goals

Per the issue scope, this PR does not:

- Delete the old dependency-state protocol (planned for #755)
- Unify top-level parsing with the same runtime model (planned for #754)
- Migrate `conditional()`, which still uses `resolveDeferredParseStates` directly

## Test plan

- [x] All 427 Deno tests pass (`mise test:deno`)
- [x] Node.js cross-runtime check passes (`mise test:node`, excluding pre-existing logtape failure)
- [x] Type check and lint clean (`mise check`)
- [x] Dependency semantics regression matrix (*dependency-semantics.test.ts*) passes unchanged
- [x] Dependency edge cases (*dependency-extra.test.ts*) pass unchanged

Closes https://github.com/dahlia/optique/issues/753

[#750]: https://github.com/dahlia/optique/issues/750
[#751]: https://github.com/dahlia/optique/issues/751
[#752]: https://github.com/dahlia/optique/issues/752
[#753]: https://github.com/dahlia/optique/issues/753
[#754]: https://github.com/dahlia/optique/issues/754
[#755]: https://github.com/dahlia/optique/issues/755

### CODE REVIEW

File: packages/core/src/dependency-runtime.ts [hunk 1]
```diff
@@ -762,3 +770,398 @@ export async function replayDerivedParserAsync(
...
+  state: Record<PropertyKey, unknown>,
+  parentPath?: readonly PropertyKey[],
+): RuntimeNode[] {
+  const prefix = parentPath ?? [];
+  const nodes: RuntimeNode[] = [];
+  for (const [field, parser] of pairs) {
+    const fieldState = Object.hasOwn(state, field) ||
+        (typeof field === "symbol" && field in state)
+      ? state[field as string | symbol]
+      : undefined;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency-runtime.ts]: Fixed in e3ad6909. `Object.hasOwn` indeed handles symbol keys, so the redundant `typeof field === "symbol" && field in state` fallback is removed.

File: packages/core/src/dependency-runtime.test.ts [hunk 2]
```diff
@@ -721,3 +727,219 @@ describe("replayDerivedParserAsync", () => {
...
+// Bridge helpers for construct migration
+// Part of https://github.com/dahlia/optique/issues/753
+// =============================================================================
+
+describe("extractRawInputFromState", () => {
+  // Create a minimal DerivedValueParser mock for DeferredParseState creation.
+  function makeDerivedVpMock() {
+    // deno-lint-ignore no-explicit-any
+    return { [Symbol.for("optique.parseWithDependency")]: () => ({}) } as any;
+  }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency-runtime.test.ts:747]: Fixed in e3ad6909. The mock now imports and uses the real `parseWithDependency` symbol constant instead of the ad-hoc `Symbol.for` string. The `as any` remains because `createDeferredParseState` expects the full `DerivedValueParser` type but we only need `parseWithDependency` and `dependencyId` for these tests.

File: packages/core/src/dependency-runtime.ts [hunk 3]
```diff
@@ -762,3 +770,398 @@ export async function replayDerivedParserAsync(
...
+
+  if (isDependencySourceState(state)) {
+    const depId = state[dependencyIdSymbol];
+    const result = state.result;
+    if (result.success && depId != null) {
+      // Always overwrite so that later values win (e.g., multiple()
+      // where the last tag value should be used as the dependency).
+      runtime.registerSource(depId, result.value, "cli");
+    }
+    return;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency-runtime.ts:886]: Fixed in e3ad6909. `collectSourcesFromState` now calls `runtime.markSourceFailed(depId)` for `success: false` sources, mirroring `collectExplicitSourceValues`.

Note that the old `collectDependencies` on `main` also silently ignored failed sources, so this is not a regression but a defense-in-depth improvement. In the current flow, the error from the failed source field is propagated by Phase 3 before any derived field result is returned, so the incorrect fallback never surfaces. But marking the failure explicitly is the right thing to do.

File: packages/core/src/dependency-runtime.ts [hunk 4]
```diff
@@ -762,3 +770,401 @@ export async function replayDerivedParserAsync(
...
+  if (isPlainObject(state)) {
+    const resolved: Record<string | symbol, unknown> = {};
+    for (const key of Reflect.ownKeys(state)) {
+      resolved[key] = resolveDeferredInState(state[key], runtime, visited);
+    }
+    return resolved;
+  }
+
+  return state;
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency-runtime.ts:976]: This matches the old `resolveDeferred()` on `main`, which uses the identical `WeakSet` pattern:

```typescript
if (visited.has(state)) return state;
visited.add(state);
```

In practice, parser states are never shared objects — each field gets its own state from parsing. The `visited` set exists only to prevent infinite loops with circular references. Switching to a memoizing `WeakMap` would be a valid improvement for correctness in edge cases, but it is out of scope for this migration since the behavior is unchanged from `main`.

File: packages/core/src/dependency-runtime.ts [hunk 5]
```diff
@@ -762,3 +770,401 @@ export async function replayDerivedParserAsync(
...
+    nodes.push({
+      path: [...prefix, i],
+      parser,
+      state: elemState,
+      matched: isMatchedState(elemState, parser),
+      defaultDependencyValues: snapshotDefaults(parser.dependencyMetadata),
+    });
+  }
+  return nodes;
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency-runtime.ts:1157]: Fixed in 41690f1. Both `buildRuntimeNodesFromPairs` and `buildRuntimeNodesFromArray` now return `readonly RuntimeNode[]`.

File: packages/core/src/constructs.ts [hunk 6]
```diff
@@ -6529,19 +6379,36 @@ type ConcatValues<TParsers extends ConcatParsers> = IsTuple<TParsers> extends
...
+    );
+    const tupleState: Record<string, unknown> = Object.fromEntries(
+      stateArray.map((s, i) => [String(i), s]),
+    );
+    completeDependencySourceDefaults(
+      { ...preParsedContext, state: tupleState } as ParserContext<unknown>,
+      pairs,
+      runtime.registry,
+      preParsedContext.exec,
+    );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Reverted in 41690f1. The `completeDependencySourceDefaults` call with top-level tuple parser pairs was indeed a no-op: `pendingDependencyDefaults` only understands leaf field parsers with `PendingDependencySourceState` markers, not composite tuple parsers.

The underlying gap (concat suggest not seeding `withDefault` defaults across children) is pre-existing on `main` — the old `buildSuggestRegistry` also only called `collectDependencies` without any default seeding. A proper fix would require extracting field-level pairs from inside each tuple child, which is a deeper change better suited for a follow-up.

File: packages/core/src/constructs.ts [hunk 7]
```diff
@@ -7010,25 +6877,33 @@ export function concat(
...
+      dependencyRuntime: runtime,
+    } as ExecutionContext;
 
-    // Phase 2: Resolve deferred parse states across all tuples
-    const resolvedCombinedState = resolveDeferredParseStates(combinedState);
+    // Phase 2: Resolve deferred parse states across all tuples using runtime.
+    const resolvedArray = resolveStateWithRuntime(
+      stateArray,
+      runtime,
+    ) as unknown[];
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:6875]: This matches the old `concat().complete()` on `main`, which also went straight from building a combined state object to `resolveDeferredParseStates(combinedState)` without any pre-completion step:

```typescript
// Phase 1: Build a combined state object for dependency resolution
const combinedState: Record<number, unknown> = {};
for (let i = 0; i < stateArray.length; i++) {
  combinedState[i] = stateArray[i];
}
// Phase 2: Resolve deferred parse states across all tuples
const resolvedCombinedState = resolveDeferredParseStates(combinedState);
```

For `withDefault` defaults in earlier children, the shared runtime propagates them when each child's `complete()` runs its own `preCompleteAndRegisterDependencies` in Phase 3. Cross-child default seeding at the concat level would be a new feature, not a regression fix.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/constructs.ts [hunk 8]
```diff
@@ -6092,37 +5946,45 @@ export function merge(
...
           const parserState = extractCompleteState(parser, resolvedState, i);
-          const result = parser.complete(
-            parserState as Parameters<typeof parser.complete>[0],
-            exec,
+          const result = unwrapCompleteResult(
+            parser.complete(
+              parserState as Parameters<typeof parser.complete>[0],
+              childExec,
+            ),
           );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:5983]: This matches the old `merge().complete()` on `main`, which also called `preCompleteAndRegisterDependencies` at the merge level and then `parser.complete()` on each child — the child's own Phase 1 re-enters and re-evaluates defaults. The double evaluation is pre-existing and not introduced by this PR.

Properly fixing it would require skipping pre-completion for sources already registered in the shared runtime. That is a meaningful behavioral change better suited for a follow-up.

File: packages/core/src/dependency-runtime.ts [hunk 9]
```diff
@@ -762,3 +770,401 @@ export async function replayDerivedParserAsync(
...
+  if (Array.isArray(state)) {
+    return state.map((item) => resolveDeferredInState(item, runtime, visited));
+  }
+
+  if (isPlainObject(state)) {
+    const resolved: Record<string | symbol, unknown> = {};
+    for (const key of Reflect.ownKeys(state)) {
+      resolved[key] = resolveDeferredInState(state[key], runtime, visited);
+    }
+    return resolved;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency-runtime.ts:972]: Fixed in 4985b73. Both sync and async `resolveDeferredInState` now use `Object.create(Object.getPrototypeOf(state))` instead of `{}` to preserve the original prototype. The old `resolveDeferred` on `main` had the same gap, but the fix is trivial and improves consistency with how *annotations.ts* and *facade.ts* handle prototype preservation.

File: packages/core/src/dependency-runtime.ts [hunk 10]
```diff
@@ -762,3 +770,401 @@ export async function replayDerivedParserAsync(
...
+  for (const [field, parser] of pairs) {
+    const fieldState = Object.hasOwn(state, field)
+      ? state[field as string | symbol]
+      : undefined;
+    nodes.push({
+      path: [...prefix, field],
+      parser,
+      state: fieldState,
+      matched: isMatchedState(fieldState, parser),
+      defaultDependencyValues: snapshotDefaults(parser.dependencyMetadata),
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/dependency-runtime.ts]: Fixed in 4985b73. Removed the eager `snapshotDefaults` call from both `buildRuntimeNodesFromPairs` and `buildRuntimeNodesFromArray`. These builders are now only used in suggest paths where `defaultDependencyValues` is never consumed (`replayDerivedParser` is not called during suggestions). The `snapshotDefaults` helper itself has been deleted as dead code.

File: packages/core/src/constructs.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:5983]: Yes, please file a separate issue, thanks!

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/constructs.ts [hunk 11]
```diff
@@ -7010,25 +6860,33 @@ export function concat(
...
+      dependencyRuntime: runtime,
+    } as ExecutionContext;
 
-    // Phase 2: Resolve deferred parse states across all tuples
-    const resolvedCombinedState = resolveDeferredParseStates(combinedState);
+    // Phase 2: Resolve deferred parse states across all tuples using runtime.
+    const resolvedArray = resolveStateWithRuntime(
+      stateArray,
+      runtime,
+    ) as unknown[];
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts:6875]: This matches the old `concat().complete()` on `main`, which also went straight from building a combined state to `resolveDeferredParseStates(combinedState)` without any pre-completion step. See the [earlier thread](https://github.com/dahlia/optique/pull/761#discussion_r3005897573) for the full explanation.

The `concat()` children are tuple parsers, and `preCompleteAndRegisterDependencies` operates on field-level pairs (option/argument parsers with dependency markers). Passing top-level tuple parsers as pairs would be ineffective since tuple parsers don't have `PendingDependencySourceState` or `wrappedDependencySourceMarker`. A proper fix would require extracting field-level pairs from inside each tuple child, which is a deeper change for a follow-up.
